rfc/system/5204-project-isolation.md

```diff
@@ -0,0 +1,110 @@
+
+## 3 Proposed Implementation
+
+The Flyte operators can enable or disable the project isolation feature. When enabled, Flyte will evaluate the identity token issued by the IDP and extract the values of a configured claim. The user can then 
+access a project if the claim value matches the project. If the project isolation feature is enabled and the claim value does not match any project, the user is denied access to all projects. So it is a whitelist mechanism.
+
+The only external interfaces of Flyte is the API that is being used by the Flyte console, the flytectl cli and pyflyte. So it is sufficient to add a check if the user interacting with the API for a certain project has the required permission.
+
+Since the API endpoints are very different from one another, the check needs to be implemented in each endpoint. Because of the different formats of the endpoint holding the project it will be hard to define a middleware centrally, which performs the check.
+
+The mapping between claim value and project to authorize can be configured in the following way and put in the Helm chart of the Flyte deployment:
+
+```yaml
+auth:
+  project-authorization:
+    enabled: true
+    claim: "entitlements"
+    mapping:
+      project1: "project1"
+      project2: "project2"
+```
+
+The identity token is evaluated by Flyte already and data can be accessed in the Flyte code. The following example golang code shows how the project isolation can be implemented in the Flyte code:
+
+
+```go
+package auth
+
+import (
+	"context"
+	"github.com/flyteorg/flyte/flyteadmin/auth/config"
+)
+
+func ProjectAccessPermitted(ctx context.Context, project string) bool {
+	authConfig := config.GetConfig()
+
+	if authConfig.ProjectAuth.isEnabled() {
+		identityContext := IdentityContextFromContext(ctx)
+
+		claimValues := identityContext.UserInfo().GetAdditionalClaims().GetFields()[authConfig.ProjectAuth.Claim].GetListValue().GetValues()
+
+		for claimValue := range claimValues {
+			if claimValue.GetStringValue() == authConfig.ProjectAuth.mapping[project] {
+				return true
+			}
+		}
+		
+		return false
+	} else {
+		return true
+	}
+
+}
+```
+
+The following piece of code also shows how the check is integrated into existing API endpoints:
+
+```go
+func (m *AdminService) GetWorkflow(ctx context.Context, request *admin.ObjectGetRequest) (*admin.Workflow, error) {
+	defer m.interceptPanic(ctx, request)
+
+    if request == nil {
+        return nil, status.Errorf(codes.InvalidArgument, "Incorrect request, nil requests not allowed")
+    }
+	
+	if ProjectAccessPermitted(ctx, request.GetId().Project) {
```

Comment:
this is an opt-in mechanism, right? how do you ensure that new endpoints also get the same treatment?

Reply:
Yes, it is an opt-in mechanism. We have meanwhile implemented it in a private fork of the project. We moved away from changing every single endpoint, but added a central middleware. But since there are no common fields for all of the request objects, actually each and every endpoint and its request object need to be explicitly mentioned, which I consider a weakness because if newly added endpoints are forgotten to be added in that middleware, then those endpoints will not benefit from the project isolation. Unfortunately we could not find a better solution to this issue.

Reply:
If you like to, I can also show you how we actually implemented it in the fork.

Reply:
@robert-ulbrich-mercedes-benz that all makes sense, I can see where maybe the middleware rejects unhandled endpoints by default but then that still requires manual code changes per new endpoint (and for newer or infrequent endpoints, it may not be obvious that the authorization check was missed)

How do you keep up with new service definition changes in your fork? Or is the process manual currently?

Reply:
We manually went over the GRPC service definitions that are available. We found out, that only Admin Services are of concern. But I share your concerns regarding new endpoints that might be missed in that central middleware.